### PR TITLE
Enhance estimate of FI memory usage

### PR DESCRIPTION
**Pull request by lancelly**
Pipelines can be organized into a tree structure. This PR is used to estimate the memory usage during the execution of a pipeline. Currently, it is assumed that in the most extreme cases, the child pipelines under a ConsumeAllChildren type pipeline will all run simultaneously. Therefore, when considering memory usage, all children must be accounted for. For ConsumeChildrenOneByOne type pipelines, only some of the child pipelines will run at the same time. Therefore, only the memory usage of these active children needs to be considered, along with the static memory occupied by the inactive children.

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/node/PlanNodeType.java [hunk 1]
```diff
@@ -209,6 +209,8 @@ public enum PlanNodeType {
   EXPLAIN_ANALYZE((short) 90),
 
   PIPE_OPERATE_SCHEMA_QUEUE_REFERENCE((short) 91),
+
+  FULL_OUTER_TIME_JOIN_REFERENCE((short) 92),
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/node/PlanNodeType.java:213]: what's this for?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LocalExecutionPlanContext.java [hunk 2]
```diff
@@ -91,6 +98,10 @@ public class LocalExecutionPlanContext {
   // use AtomicReference not for thread-safe, just for updating same field in different pipeline
   private AtomicReference<List<Long>> timePartitions = new AtomicReference<>();
 
+  /** Records the parent of each pipeline. The order of each list does not matter for now. */
+  private Map<PlanNodeId, List<PipelineMemoryEstimator>> parentPlanNodeIdToMemoryEstimator =
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LocalExecutionPlanContext.java:102]: remember to set this map to null, after you check the memory, this Map may occupy huge memory when there are many devices;

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/node/PlanNodeType.java (hunk 1 above)
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/node/PlanNodeType.java]: done

Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/node/PlanNodeType.java]: Done.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LocalExecutionPlanContext.java (hunk 2 above)
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LocalExecutionPlanContext.java:102]: Done.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/OperatorTreeGenerator.java [hunk 3]
```diff
@@ -2947,6 +2949,18 @@ public List<Operator> dealWithConsumeAllChildrenPipelineBreaker(
...
+        // the parent of the pipeline of ScanNode should be IdentitySinkNode in the map, otherwise
+        // we will lose the information of these pipelines
+        List<PipelineMemoryEstimator> childPipelineMemoryEstimators =
+            context.getParentPlanNodeIdToMemoryEstimator().get(localChild.getPlanNodeId());
+        if (childPipelineMemoryEstimators != null) {
+          context.getParentPlanNodeIdToMemoryEstimator().remove(localChild.getPlanNodeId());
+          context
+              .getParentPlanNodeIdToMemoryEstimator()
+              .put(node.getPlanNodeId(), childPipelineMemoryEstimators);
+        }
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/OperatorTreeGenerator.java]: I updated the comment.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/LastCacheScanOperator.java [hunk 4]
```diff
@@ -83,4 +87,11 @@ public long calculateRetainedSizeAfterCallingNext() {
   public PlanNodeId getSourceId() {
     return sourceId;
   }
+
+  @Override
+  public long getEstimatedMemoryUsageInBytes() {
+    return INSTANCE_SIZE
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/LastCacheScanOperator.java:93]: Done.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/last/UpdateLastCacheOperator.java [hunk 5]
```diff
@@ -89,4 +94,14 @@ protected void appendLastValueToTsBlockBuilder(long lastTime, TsPrimitiveType la
     LastQueryUtil.appendLastValue(
         tsBlockBuilder, lastTime, fullPath.getFullPath(), lastValue.getStringValue(), dataType);
   }
+
+  @Override
+  public long getEstimatedMemoryUsageInBytes() {
+    return INSTANCE_SIZE
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/last/UpdateLastCacheOperator.java:100]: Now I refactored EstimatedSizeOfPartialPathWithoutClassSize to EstimatedSizeOfPartialPath in which the INSTANCE_SIZE of PartialPath is calculated. I removed INSTANCE_SIZE of PartialPath in other classes.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/last/LastQueryTransformOperator.java [hunk 6]
```diff
@@ -117,4 +122,13 @@ public void close() throws Exception {
     }
     tsBlockBuilder = null;
   }
+
+  @Override
+  public long getEstimatedMemoryUsageInBytes() {
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/last/LastQueryTransformOperator.java]: Done.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/last/LastQueryOperator.java [hunk 7]
```diff
@@ -173,4 +178,15 @@ public long calculateRetainedSizeAfterCallingNext() {
     }
     return max;
   }
+
+  @Override
+  public long getEstimatedMemoryUsageInBytes() {
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/last/LastQueryOperator.java]: Done. I added the size of tsBlockBuilder in all operators that have this field.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/MemoryMeasurable.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.queryengine.execution;
+
+public interface MemoryMeasurable {
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/MemoryMeasurable.java]: Changed to Accountable.

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**